### ISSUE TITLE
real life metric measurements with Ruicheng/moge-2-vitl-normal

### ISSUE THREAD

**Issue by Gabriel-Wainmann-SnapTradie** [microsoft/MoGe#75]
Hi

Getting the point cloud result seems to be larger (2x? more? less?) than real-life distances.

What am I doing wrong?

Thank you!

`"""
Google Colab script for video to point cloud conversion using MoGe

Cell 1 - Install dependencies:
!pip install torch torchvision git+https://github.com/microsoft/MoGe.git opencv-contrib-python open3d

Cell 2 - Main script:
"""

# Imports
import cv2
import numpy as np
import open3d as o3d
import torch
from moge.model.v2 import MoGeModel
from google.colab import files
import time
import os
import shutil
import threading
from queue import Queue
from dataclasses import dataclass
from typing import Tuple

@dataclass
class FrameData:
    """Container for frame data through the pipeline"""
    index: int
    original: np.ndarray
    processed: np.ndarray = None
    tensor: torch.Tensor = None

# Load model
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Using device: {device}")
model = MoGeModel.from_pretrained("Ruicheng/moge-2-vitl-normal").to(device).half().eval()

# Create/clean output folder
output_folder = 'point_clouds'
if os.path.exists(output_folder):
    shutil.rmtree(output_folder)
os.makedirs(output_folder)

# Video
video_path = 'Joe.mp4'

# Parameters
frame_interval_sec = 0.5  # Interval between frames to process (seconds)
scale_percent = 50        # Percentage to downscale frames
min_distance = 1.5       # Minimum distance from camera (meters)

# Process video
cap = cv2.VideoCapture(video_path)
fps = cap.get(cv2.CAP_PROP_FPS)
frame_skip = int(fps * frame_interval_sec)
width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

# Camera calibration parameters (for 4K: 3840×2160)
fx = 2191.59764
fy = 2185.196
cx = 1901.801306
cy = 1039.412144

camera_matrix = np.array([[fx, 0, cx],
                         [0, fy, cy],
                         [0, 0, 1]], dtype=float)

dist_coeffs = np.array([-0.06877963, 0.08537744, -0.00089973, -0.00070357, -0.01667835], dtype=float)

# Set up queues
gpu_queue = Queue(maxsize=8)
write_queue = Queue()

def preprocess_frame(frame_data: FrameData, scale_percent: int, camera_matrix: np.ndarray, dist_coeffs: np.ndarray) -> FrameData:
    """Preprocess a single frame: resize, undistort, and convert to tensor"""
    frame = frame_data.original
    
    # Downscale if needed
    if scale_percent != 100:
        scale = scale_percent / 100
        frame = cv2.resize(frame, (int(width*scale), int(height*scale)))
        scaled_camera_matrix = camera_matrix.copy()
        scaled_camera_matrix[0:2, :] *= scale
    else:
        scaled_camera_matrix = camera_matrix
    
    # Undistort
    frame = cv2.undistort(frame, scaled_camera_matrix, dist_coeffs)
    
    # Convert BGR to RGB
    rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    
    # Create tensor
    tensor = torch.from_numpy(rgb/255.0).to(device, dtype=torch.float16).permute(2,0,1)
    
    frame_data.processed = rgb
    frame_data.tensor = tensor
    return frame_data

def filter_points_gpu(points_gpu: torch.Tensor, mask_gpu: torch.Tensor, rgb_tensor: torch.Tensor, 
                     min_distance: float = 1.5) -> Tuple[np.ndarray, np.ndarray]:
    """Filter point cloud data on GPU before transferring to CPU"""
    # Calculate Euclidean distance from camera
    distance = torch.sqrt(points_gpu[..., 0]**2 + points_gpu[..., 1]**2 + points_gpu[..., 2]**2)
    
    # Filter points based on:
    # 1. MoGe's confidence mask
    # 2. Minimum distance from camera
    # 3. Valid (finite) values
    valid_mask = mask_gpu & (distance > min_distance) & torch.isfinite(points_gpu).all(dim=-1)
    
    # Flatten and extract valid points on GPU
    h, w = mask_gpu.shape
    points_flat = points_gpu.view(-1, 3)
    rgb_flat = rgb_tensor.view(-1, 3)
    mask_flat = valid_mask.view(-1)
    
    valid_points_gpu = points_flat[mask_flat]
    valid_colors_gpu = rgb_flat[mask_flat]
    
    # Transfer only filtered points to CPU
    valid_points = valid_points_gpu.cpu().numpy()
    valid_colors = valid_colors_gpu.cpu().numpy()
    
    return valid_points, valid_colors

def writer_thread():
    """Background thread for writing PLY files"""
    while True:
        item = write_queue.get()
        if item is None:
            break
        
        filename, points, colors = item
        try:
            # Create Open3D object here (in background thread)
            pcd = o3d.geometry.PointCloud()
            pcd.points = o3d.utility.Vector3dVector(points)
            pcd.colors = o3d.utility.Vector3dVector(colors / 255.0)
            
            # Write PLY file
            o3d.io.write_point_cloud(filename, pcd, write_ascii=False)
        except Exception as e:
            print(f"Error writing {filename}: {e}")
        finally:
            write_queue.task_done()

def gpu_inference_thread():
    """GPU inference thread - process one frame at a time"""
    frame_count = 0
    total_gpu_time = 0
    total_time = 0
    
    while True:
        frame_data = gpu_queue.get()
        
        if frame_data is None:  # Sentinel
            gpu_queue.task_done()
            break
        
        # Process single frame
        start_time = time.time()
        
        # GPU inference
        with torch.no_grad():
            outputs = model.infer(frame_data.tensor)
        
        # Get RGB data from input tensor (already on GPU)
        rgb_gpu = frame_data.tensor.permute(1, 2, 0) * 255.0
        
        # Filter on GPU before CPU transfer
        valid_points, valid_colors = filter_points_gpu(
            outputs['points'], 
            outputs['mask'], 
            rgb_gpu, 
            min_distance
        )
        
        # Clean GPU memory
        torch.cuda.empty_cache()
        
        gpu_time = time.time() - start_time
        
        # Send filtered data to writer
        filename = os.path.join(output_folder, f'frame_{frame_data.index:04d}.ply')
        write_queue.put((filename, valid_points, valid_colors))
        
        total_frame_time = time.time() - start_time
        
        # Update statistics
        frame_count += 1
        total_gpu_time += gpu_time
        total_time += total_frame_time
        
        # Print progress every 10 frames
        if frame_count % 10 == 0:
            avg_gpu = total_gpu_time / frame_count
            avg_total = total_time / frame_count
            print(f"Processed {frame_count} frames: Avg GPU {avg_gpu:.3f}s, Avg Total {avg_total:.3f}s per frame")
            
            # Also show current GPU memory usage
            if torch.cuda.is_available():
                allocated = torch.cuda.memory_allocated() / 1024**3
                reserved = torch.cuda.memory_reserved() / 1024**3
                print(f"  GPU Memory: {allocated:.2f}GB allocated, {reserved:.2f}GB reserved")
        
        gpu_queue.task_done()
    
    # Final statistics
    if frame_count > 0:
        print(f"\nGPU thread complete: {frame_count} frames")
        print(f"Average GPU time: {total_gpu_time/frame_count:.3f}s per frame")
        print(f"Average total time: {total_time/frame_count:.3f}s per frame")
    
    # Signal end to write queue
    write_queue.put(None)

# Start threads
writer = threading.Thread(target=writer_thread, daemon=True)
gpu_thread = threading.Thread(target=gpu_inference_thread, daemon=True)

writer.start()
gpu_thread.start()

# Main loop - read video and preprocess frames
frame_idx = 0
processed_count = 0
start_total = time.time()

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break
    
    if frame_idx % frame_skip == 0:
        # Create frame data
        frame_data = FrameData(index=frame_idx, original=frame)
        
        # Preprocess frame
        preprocess_frame(frame_data, scale_percent, camera_matrix, dist_coeffs)
        
        # Send to GPU queue
        gpu_queue.put(frame_data)
        processed_count += 1
    
    frame_idx += 1

# Signal end of stream
gpu_queue.put(None)

# Wait for completion
gpu_thread.join()
writer.join()

# Statistics
total_time = time.time() - start_total
print(f"\nProcessed {processed_count} frames")
print(f"Total time: {total_time:.2f}s")
print(f"Average time per frame: {total_time/processed_count:.3f}s")
print(f"Overall FPS: {processed_count/total_time:.2f}")

cap.release()
print("\nAll files written successfully!")

# Download results (add to a separate cell in Colab)
# !zip -r point_clouds.zip point_clouds/
# files.download('point_clouds.zip')
`

**Comment by EasternJournalist** [comment]
Hi. It looks like you're interpreting the predicted intrinsics as if they were in pixel space, which leads to an incorrect FOV computation—nearly 180 degrees.

You're currently using:

```
fov_x_rad = 2 * np.arctan(w / (2 * fx))
fov_y_rad = 2 * np.arctan(h / (2 * fy))
```

However, since the output fx and fy are already normalized by image width and height respectively, the correct computation should be:

```
fov_x_rad = 2 * np.arctan(1 / (2 * fx))
fov_y_rad = 2 * np.arctan(1 / (2 * fy))
```

This should give you the correct field of view in radians.

Regarding the real-life metric scale: could you share some examples of your input images? In general, the model produces the most accurate metric scales for typical indoor and street scenes. For unfamiliar objects or extremely close-up views, the predictions may be less reliable due to limited training data in those domains.

**Comment by Gabriel-Wainmann-SnapTradie** [comment]
Thank you for this. I'll check the intrinsics aspect.

As to the depth - please see attached a photo.
All the distances come out about double their real distance from the camera. The end of the lance is about 75 cm from the camera, for example.
 
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/699bbdd5-7f4d-4a34-b6ed-5772c60bbd69" />

**Field-of-View (FoV)**
Diagonal 93°, Horizontal 81°

**Camera matrix**
 [[1.09579882e+03 0.00000000e+00 9.50900653e+02]
 [0.00000000e+00 1.09259800e+03 5.19706072e+02]
 [0.00000000e+00 0.00000000e+00 1.00000000e+00]]

**Distortion coeffs**
[[-0.06877963  0.08537744 -0.00089973 -0.00070357 -0.01667835]]

**Comment by sunshineywz123** [comment]
![Image](https://github.com/user-attachments/assets/779141f5-e50d-4a41-bcf4-5a9e83da388f)
My scene has a car, but when I compare it using moge2 and lidar, there is still a significant difference in scale.

**Comment by sunshineywz123** [comment]
my gt intrins as input

**Comment by Tetsujinfr** [comment]
Hi, great model and great repository too, have a blast with this for over a year now !

W.r.t this trail topic on distances estimation I did a test with two images in the wild to check estimation versus inter-pupillary distance which is known to be around 7cm for adults (can vary a bit but not much). Another measure I could have used is the distance between the hears or the size of the mouth etc: they can vary a bit but I think those are good reference points for a model to use, to learn metric distances provided there are humans within the training dataset of course.

So I have one close up image of Miyazaki on one side, and another image of Musk standing next to a car (my "common object" choice for the test)
Results images below. 

Observations:
Indeed the estimate of the inter-pupillary distance for the Musk image with the car is really good with 7cm.
And indeed the estimate for Miyazaki (without any car or common abject) is bad but **TWICE** the value of a good estimate.

So, I conclude that @EasternJournalist you did not use human body standard distances to train the model, correct? (would have been so good to use those though)

Based on @Gabriel-Wainmann-SnapTradie 's observation above, and on mine above too, I appreciate only 2 examples here, it seems that when no "common object" is in the image the estimated distances are twice as big as good estimates: @EasternJournalist is it something you have observed too and do you think the model could still have learned the right distances X 2 somehow? 

thanks a lot

![Image](https://github.com/user-attachments/assets/a77cd914-a08b-4028-8005-e7f944ba39df)

![Image](https://github.com/user-attachments/assets/8c44c3d8-73d8-45d3-90c8-56956f52e002)
